Mathlib/Tactic/CC.lean

```diff
@@ -0,0 +1,286 @@
+#align cc_state.proof_for_false Mathlib.Tactic.CC.CCState.proofForFalse
+
+#align cc_state.mk Mathlib.Tactic.CC.CCState.mk
+
+/-- Create a congruence closure state object using the hypotheses in the current goal. -/
+def mkUsingHs : MetaM CCState :=
+  CCState.mkUsingHsCore {}
+#align cc_state.mk_using_hs Mathlib.Tactic.CC.CCState.mkUsingHs
+
+/-- The root expressions for each equivalence class in the graph. -/
+def roots (s : CCState) : List Expr :=
+  CCState.rootsCore s true
+#align cc_state.roots Mathlib.Tactic.CC.CCState.roots
+
+instance : ToMessageData CCState :=
+  ⟨fun s => CCState.ppEqcs s true⟩
+
+/-- Continue to append following expressions in the equivalence class of `e` to `r` until `f` is
+found. -/
+partial def eqcOfCore (s : CCState) (e : Expr) (f : Expr) (r : List Expr) : List Expr :=
+  let n := s.next e
+  if n == f then e :: r else eqcOfCore s n f (e :: r)
+#align cc_state.eqc_of_core Mathlib.Tactic.CC.CCState.eqcOfCore
+
+/-- The equivalence class of `e`. -/
+def eqcOf (s : CCState) (e : Expr) : List Expr :=
+  s.eqcOfCore e e []
+#align cc_state.eqc_of Mathlib.Tactic.CC.CCState.eqcOf
+
+/-- The size of the equivalence class of `e`. -/
+def eqcSize (s : CCState) (e : Expr) : Nat :=
+  s.eqcOf e |>.length
+#align cc_state.eqc_size Mathlib.Tactic.CC.CCState.eqcSize
+
+/-- Fold `f` over the equivalence class of `c`, accumulating the result in `a`.
+Loops until the element `first` is encountered.
+
+See `foldEqc` for folding `f` over all elements of the equivalence class. -/
+partial def foldEqcCore {α} (s : CCState) (f : α → Expr → α) (first : Expr) (c : Expr) (a : α) :
+    α :=
+  let new_a := f a c
+  let next := s.next c
+  if next == first then new_a else foldEqcCore s f first next new_a
+#align cc_state.fold_eqc_core Mathlib.Tactic.CC.CCState.foldEqcCore
+
+/-- Fold the function of `f` over the equivalence class of `e`. -/
+def foldEqc {α} (s : CCState) (e : Expr) (a : α) (f : α → Expr → α) : α :=
+  foldEqcCore s f e e a
+#align cc_state.fold_eqc Mathlib.Tactic.CC.CCState.foldEqc
+
+/-- Fold the monadic function of `f` over the equivalence class of `e`. -/
+def foldEqcM {α} {m : Type → Type} [Monad m] (s : CCState) (e : Expr) (a : α)
+    (f : α → Expr → m α) : m α :=
+  foldEqc s e (return a) fun act e => do
+    let a ← act
+    f a e
+#align cc_state.mfold_eqc Mathlib.Tactic.CC.CCState.foldEqcM
+
+end CCState
+
+/--
+The congruence closure tactic `cc` tries to solve the goal by chaining
+equalities from context and applying congruence (i.e. if `a = b`, then `f a = f b`).
+It is a finishing tactic, i.e. it is meant to close
+the current goal, not to make some inconclusive progress.
+A mostly trivial example would be:
+
+```lean
+example (a b c : ℕ) (f : ℕ → ℕ) (h: a = b) (h' : b = c) : f a = f c := by
+  cc
+```
+
+As an example requiring some thinking to do by hand, consider:
+
+```lean
+example (f : ℕ → ℕ) (x : ℕ)
+    (H1 : f (f (f x)) = x) (H2 : f (f (f (f (f x)))) = x) :
+    f x = x := by
+  cc
+``` -/
```

Comment:
Ah, with my second suggestion I meant that this comment could be something like:
````suggestion
/--
Applies congruence closure to solve the given metavariable.
This procedure tries to solve the goal by chaining
equalities from context and applying congruence (i.e. if `a = b`, then `f a = f b`).

The tactic works by building an equality matching graph. It's a graph where
the vertices are terms and they are linked by edges if they are known to
be equal. Once you've added all the equalities in your context, you take
the transitive closure of the graph and, for each connected component
(i.e. equivalence class) you can elect a term that will represent the
whole class and store proofs that the other elements are equal to it.
You then take the transitive closure of these equalities under the
congruence lemmas.
The `cc` implementation in Lean does a few more tricks: for example it
derives `a = b` from `Nat.succ a = Nat.succ b`, and `Nat.succ a != Nat.zero` for any `a`.
* The starting reference point is Nelson, Oppen, [Fast decision procedures based on congruence
closure](http://www.cs.colorado.edu/~bec/courses/csci5535-s09/reading/nelson-oppen-congruence.pdf),
Journal of the ACM (1980)
* The congruence lemmas for dependent type theory as used in Lean are described in
[Congruence closure in intensional type theory](https://leanprover.github.io/papers/congr.pdf)
(de Moura, Selsam IJCAR 2016).
-/
````

That way it wouldn't duplicate the `cc` elaborator, while still providing important information. Does that work for you?

Reply:
Good.